core/miner/storage_fsm/impl/precommit_batcher_impl.cpp

```diff
@@ -0,0 +1,137 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "miner/storage_fsm/impl/precommit_batcher_impl.hpp"
+
+#include <utility>
+
+namespace fc::mining {
+
+  PreCommitBatcherImpl::PreCommitBatcherImpl(size_t maxTime,
+                                             std::shared_ptr<FullNodeApi> api)
+      : api_(std::move(api)){};
+
+  PreCommitBatcherImpl::preCommitEntry::preCommitEntry(
+      primitives::TokenAmount number, api::SectorPreCommitInfo info)
+      : deposit(std::move(number)), pci(std::move(info)){};
+
+  PreCommitBatcherImpl::preCommitEntry &
+  PreCommitBatcherImpl::preCommitEntry::operator=(const preCommitEntry &x) =
+      default;
+
+  outcome::result<std::shared_ptr<PreCommitBatcherImpl>>
+  PreCommitBatcherImpl::makeBatcher(
+      size_t maxWait,
+      std::shared_ptr<FullNodeApi> api,
+      std::shared_ptr<libp2p::protocol::Scheduler> scheduler,
+      Address &miner_address) {
+    struct make_unique_enabler : public PreCommitBatcherImpl {
+      make_unique_enabler(size_t MaxTime, std::shared_ptr<FullNodeApi> api)
+          : PreCommitBatcherImpl(MaxTime, std::move(api)){};
+    };
+
+    std::shared_ptr<PreCommitBatcherImpl> batcher =
+        std::make_unique<make_unique_enabler>(maxWait, std::move(api));
+
+    batcher->miner_address_ = miner_address;
+    batcher->maxDelay = maxWait;
+
+    batcher->cutoffStart = std::chrono::system_clock::now();
+    batcher->closestCutoff = maxWait;
+    batcher->handle_ = scheduler->schedule(maxWait, [=]() {
+      auto maybe_send = batcher->sendBatch();
```

Comment:
You cannot call outcome::result functions like this.
```suggestion
      OUTCOME_TRY(batcher->sendBatch());
```

Reply:
In this case you need to change `scheduler->schedule(maxWait, [=]()` to
`scheduler->schedule(maxWait, [=]() -> outcome::result<void>` or something like that.
Unfortunately, I cannot see libp2p::protocol::scheduler::Handle signature to suggest a good variant.
Probably you need to check `maybe_send.has_error()` instead of OUTCOME_TRY

Reply:
As the sendBatched will be called with delay, I think now the only option is to add results to logs to get info about it

Reply:
Agree